File: server/src/main/java/com/vgolos/VGolos/service/CitizenService.java
package com.vgolos.VGolos.service;

import com.vgolos.VGolos.entity.Citizen;
import com.vgolos.VGolos.repository.CitizenRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.LinkedList;
import java.util.List;
import java.util.Date;
import java.util.concurrent.TimeUnit;

@Service
public class CitizenService {

    private CitizenRepository citizenRepository;

    @Autowired
    public CitizenService(CitizenRepository citizenRepository) {
        this.citizenRepository = citizenRepository;
    }

    public List<Citizen> findAll() {
        return citizenRepository.findAll();
    }

    public Citizen findByLogin(String login) {
        return citizenRepository.findByLogin(login).get();
    }

    public Citizen findByIdn(String idn) {
        return citizenRepository.findByIdn(idn).get();
    }

    public Citizen findById(Long id) {
        return citizenRepository.findById(id).get();
    }

    public long getAgeInDays(Long id){
        Date currentDate = new Date();
        Long diffInMillies = Math.abs(currentDate.getTime() - citizenRepository.findById(id).get().getDateOfBirth().getTime());
        return TimeUnit.DAYS.convert(diffInMillies, TimeUnit.MILLISECONDS);
    }

    public long getAge(Long id){
        Date currentDate = new Date();
        Long diffInMillies = Math.abs(currentDate.getTime() - citizenRepository.findById(id).get().getDateOfBirth().getTime());
        return TimeUnit.DAYS.convert(diffInMillies, TimeUnit.MILLISECONDS) / 365;
    }

    public List<Citizen> search(String name) {
       return citizenRepository.findByFirstNameContainingIgnoreCaseOrLastNameContainingIgnoreCase(name, name);
    }

    public Citizen createCitizen(Citizen citizen) {
        citizen.setId(-1L);
        citizen.setLogin(citizen.getIdn());
        Citizen createdCitizen = citizenRepository.save(citizen);
        return createdCitizen;
    }

    public Citizen update(Citizen account) {
        Citizen citizen = citizenRepository.findById(account.getId()).get();
        citizen.setRole(account.getRole());
        citizen.setFirstName(account.getFirstName());
        citizen.setLastName(account.getLastName());
        citizen.setRegion(account.getRegion());
        return citizenRepository.save(citizen);
    }

    public void deleteById(Long id) {
        citizenRepository.deleteById(id);
    }

    public boolean isAdult(Long id){
        if (getAgeInDays(id) >= 6553){
            return true;
        }
        else return false;
    }
}

File: client/src/app/service/citizen.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Citizen } from '../model/citizen';
import { Empty } from '../model/empty';



@Injectable({
  providedIn: 'root'
})
export class CitizenService {
  constructor(private http: HttpClient) { }
  baseUrl: string = 'http://localhost:8080/citizens/';

  findAll(): Observable<Citizen[]> {
    return this.http.get<Citizen[]>(this.baseUrl);
  }

  findByLogin(login:string):Observable<Citizen>{
    return this.http.get<Citizen>(this.baseUrl+"login/" + login);
  }

  findByIdn(idn:string):Observable<Citizen> {
    return this.http.get<Citizen>(this.baseUrl + "idn/" + idn);
  }

  create(citizen:Citizen):Observable<Citizen> {
    return this.http.post<Citizen>(this.baseUrl+ "create", citizen);
  }

  update(citizen:Citizen): Observable<Citizen>{
    return this.http.put<Citizen>(this.baseUrl + "update", citizen);
  }

  delete(id: number): Observable<Empty> {
    return this.http.delete<Empty>(this.baseUrl + "delete/" + id);
  }

  findbyid(id:number): Observable<Citizen>{
    return this.http.get<Citizen>(this.baseUrl+ "id/" + id);
  }

  getAgeInDays(id:number): Observable<number>{
    return this.http.get<number>(this.baseUrl+"getAgeInDays/" + id);
  }

  getAge(id:number): Observable<number>{
    return this.http.get<number>(this.baseUrl+"getAge/" + id);
  }

  search(name:string): Observable<Citizen[]>{
    return this.http.get<Citizen[]>(this.baseUrl+"search/" + name);
  }

  isAdult(id:number): Observable<boolean>{
    return this.http.get<boolean>(this.baseUrl+"isAdult/" + id);
  }
  
}
